### PR TITLE
Adding tracing instrumentations in codesites

### PR DESCRIPTION
**Pull request by tanjialiang**
Hooking up tracing logic with code sites. Tracing started on query post if both system and user enables the tracing.
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 1]
```diff
@@ -1127,7 +1127,7 @@ public SystemSessionProperties(
                 booleanProperty(
                         ENABLE_DISTRIBUTED_TRACING,
                         "Enable distributed tracing of the query",
-                        tracingConfig.getEnableDistributedTracing(),
+                        false,
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1130]: Why this change?

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java [hunk 2]
```diff
@@ -244,6 +245,13 @@ public boolean isSlugValid(String slug)
         return this.slug.equals(slug);
     }
 
+    public Tracer getTracer()
+    {
+        Optional<Tracer> tracer = this.session.getTracer();
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:250]: remove 'this.'

File: presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java [hunk 3]
```diff
@@ -134,6 +137,7 @@ public QueuedStatementResource(
 
         this.responseExecutor = requireNonNull(executor, "responseExecutor is null").getExecutor();
         this.timeoutExecutor = requireNonNull(executor, "timeoutExecutor is null").getScheduledExecutor();
+        this.tracerProvider = tracerProvider;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java:140]: requireNonNull

File: presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java [hunk 4]
```diff
@@ -170,7 +174,7 @@ public Response postStatement(
             throw badRequest(BAD_REQUEST, "SQL statement is empty");
         }
 
-        SessionContext sessionContext = new HttpRequestSessionContext(servletRequest, sqlParserOptions);
+        SessionContext sessionContext = new HttpRequestSessionContext(servletRequest, sqlParserOptions, tracerProvider.getNewTracer());
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java:177]: - Does tracer starts its timer when it is created? The reason for this question is that if it doesn't start when it is created, then we don't have to pass it in here... If it is, I agree we need to initialize it as early as possible. It would be good to add a comment to explain.
- Can we add a TODO to pass in `traceToken` into `getNewTracer`?

File: presto-main/src/main/java/com/facebook/presto/server/protocol/ExecutingStatementResource.java [hunk 5]
```diff
@@ -100,7 +101,15 @@ public void getQueryResults(
...
         ListenableFuture<Response> queryResultsFuture = transform(
                 query.waitForResults(token, uriInfo, proto, wait, targetResultSize),
-                results -> toResponse(query, results, compressionEnabled),
+                (results) -> {
+                    Tracer tracer = query.getTracer();
+                    tracer.addPoint("Sending result data.");
+                    Response response = toResponse(query, results, compressionEnabled);
+                    if (results.getNextUri() == null) {
+                        tracer.endTrace("All results sent, query finished.");
+                    }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/ExecutingStatementResource.java:110]: We will have a lot of such points per query. Is that intended?

File: presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java [hunk 6]
```diff
@@ -484,7 +488,10 @@ public synchronized QueryResults getInitialQueryResults(UriInfo uriInfo, String
             // Hence it is safe to hardcode the token to be 0.
             return transform(
                     query.waitForResults(0, uriInfo, getScheme(xForwardedProto, uriInfo), maxWait, TARGET_RESULT_SIZE),
-                    results -> QueryResourceUtil.toResponse(query, results, compressionEnabled),
+                    results -> {
+                        query.getTracer().addPoint("Result is ready, redirecting to execution endpoint to fetch result data");
+                        return QueryResourceUtil.toResponse(query, results, compressionEnabled);
+                    },
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java:494]: Same as the question above. Do we need to track all HTTP requests (which could be thousands of them) back to client?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 1 above)
Comment (tanjialiang) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1130]: Because the system set session here does not really make sense. The session property set by users is used at the very beginning at the post statement, where even set by default will not be seen at that stage. So if the default of the enable_distributed_tracing is true here, the system will still not trace at the post stage unless the user client explicitly set the session to be true. 
This will bring up a situation where users check session properties and sees this one default to true, and start a query but the query is not actually traced.

I used this tracing config property at system startup to determine which tracer to use based on if tracing is enabled.

I am not sure if I explained it clearly but let me know and we can VC to clarify.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java (hunk 4 above)
Comment (tanjialiang) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java:177]: Yes the trace starts its timer when it is created (by calling TracerProvider.getNewTracer()). I'll add a TODO here.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/ExecutingStatementResource.java (hunk 5 above)
Comment (tanjialiang) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/ExecutingStatementResource.java:110]: I am not sure. But will be good to have a discussion with you as you know this flow better. I thought this was a good checkpoint when query approaches to finishing. But let's discuss it.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 1 above)
Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:1130]:
```java
SessionContext sessionContext = new HttpRequestSessionContext(servletRequest, sqlParserOptions, tracer, sessionProperties);

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 7]
```diff
@@ -180,6 +182,7 @@ else if (nameParts.size() == 2) {
             }
         }
         this.systemProperties = systemProperties.build();
+        this.sessionPropertyManager = sessionPropertyManager;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:185]: Move this after tracer's assignment

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 8]
```diff
@@ -463,8 +466,25 @@ public boolean supportClientTransaction()
 
     private boolean isTracingEnabled()
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:467]: add javadoc to explain why we need such complication

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 9]
```diff
@@ -463,8 +466,25 @@ public boolean supportClientTransaction()
...
     {
-        return systemProperties.getOrDefault(SystemSessionProperties.ENABLE_DISTRIBUTED_TRACING, "")
-                .equalsIgnoreCase("true");
+        String clientValue = systemProperties.getOrDefault(SystemSessionProperties.ENABLE_DISTRIBUTED_TRACING, "");
+
+        // Client session setting overrides everything.
+        if (clientValue.equalsIgnoreCase("true")) {
+            return true;
+        }
+        else if (clientValue.equalsIgnoreCase("false")) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:475]: `else ` is redundant

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 10]
```diff
@@ -463,8 +466,25 @@ public boolean supportClientTransaction()
...
+            return false;
+        }
+        else {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:478]: remove `else`

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 11]
```diff
@@ -463,8 +466,25 @@ public boolean supportClientTransaction()
 
     private boolean isTracingEnabled()
     {
-        return systemProperties.getOrDefault(SystemSessionProperties.ENABLE_DISTRIBUTED_TRACING, "")
-                .equalsIgnoreCase("true");
+        String clientValue = systemProperties.getOrDefault(SystemSessionProperties.ENABLE_DISTRIBUTED_TRACING, "");
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:469]: static import ENABLE_DISTRIBUTED_TRACING

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 12]
```diff
@@ -463,8 +466,25 @@ public boolean supportClientTransaction()
...
+        else {
+            if (sessionPropertyManager.isPresent()) {
+                // Client not set, we then take system default value
+                return sessionPropertyManager.get().decodeSystemPropertyValue(SystemSessionProperties.ENABLE_DISTRIBUTED_TRACING, null, Boolean.class);
+            }
+            else {
+                // For all other scenarios we assume tracing is disabled
+                return false;
+            }
+        }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:487]: nit
```java
return sessionPropertyManager.map(manager -> manager.decodeSystemPropertyValue(ENABLE_DISTRIBUTED_TRACING, null, Boolean.class)).orElse(false);

File: presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java [hunk 13]
```diff
@@ -134,6 +140,8 @@ public QueuedStatementResource(
 
         this.responseExecutor = requireNonNull(executor, "responseExecutor is null").getExecutor();
         this.timeoutExecutor = requireNonNull(executor, "timeoutExecutor is null").getScheduledExecutor();
+        this.tracerProvider = requireNonNull(tracerProvider, "tracerProvider is null");
+        this.sessionPropertyManager = sessionPropertyManager;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java:144]: requireNonNull

File: presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.execution.StateMachine;
+import com.facebook.presto.spi.tracing.Tracer;
+
+import static java.util.Objects.requireNonNull;
+
+public class QueryStateTracingListener
+        implements StateMachine.StateChangeListener<QueryState>
+{
+    private static final Logger log = Logger.get(QueryStateTracingListener.class);
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java:26]: not used

File: presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+                // Initial state condition
+                tracer.startBlock(newState.toString(), "");
+                prevState = newState;
+                return;
+            }
+            tracer.endBlock(prevState.toString(), "");
+            if (newState.isDone()) {
+                // Last state
+                prevState = newState;
+                System.out.println(tracer.toString());
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java:49]: remove

File: presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+import com.facebook.presto.execution.StateMachine;
+import com.facebook.presto.spi.tracing.Tracer;
+
+import static java.util.Objects.requireNonNull;
+
+public class QueryStateTracingListener
+        implements StateMachine.StateChangeListener<QueryState>
+{
+    private static final Logger log = Logger.get(QueryStateTracingListener.class);
+    private QueryState prevState;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java:27]: `previousState`

File: presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+                tracer.startBlock(newState.toString(), "");
+                prevState = newState;
+                return;
+            }
+            tracer.endBlock(prevState.toString(), "");
+            if (newState.isDone()) {
+                // Last state
+                prevState = newState;
+                System.out.println(tracer.toString());
+                tracer.endTrace("Query finished with state " + newState.toString());
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java:50]: `.toString()` is redundant

File: presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+                return;
+            }
+            else {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java:53]: remove `else`

File: presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    public void stateChanged(QueryState newState)
+    {
+        synchronized (tracer) {
+            if (prevState == null) {
+                // Initial state condition
+                tracer.startBlock(newState.toString(), "");
+                prevState = newState;
+                return;
+            }
+            tracer.endBlock(prevState.toString(), "");
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java:45]: If the previous `if` branch is to return, then we'd better break an empty line after `}` to denote it is separate logic. Same for the one below.

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java (hunk 7 above)
Comment (tanjialiang) [reply presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:185]: We need this assignment before tracer's to tell if tracing is enabled.

Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:185]: Then let's
- move the declaration of `private final Optional<SessionPropertyManager> sessionPropertyManager;` above `tracer`. 
- move `sessionPropertyManager` also right above `tracer`

Overall, the assignment in constructor should follow the same order as the declaration on stack

File: presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java (hunk 15 above)
Comment (tanjialiang) [reply presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java:49]: Ah.. slips through..

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java (hunk 7 above)
Comment (tanjialiang) [reply presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:185]: I can move this closer to tracer though

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 20]
```diff
@@ -112,19 +113,23 @@
     private final Map<SqlFunctionId, SqlInvokedFunction> sessionFunctions;
 
     private final Optional<Tracer> tracer;
+    private final Optional<SessionPropertyManager> sessionPropertyManager;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:116]: swap the positions between `tracer` and `sessionPropertyManager` because `sessionPropertyManager` is assigned earlier than `tracer` in the constructor.

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 21]
```diff
@@ -461,10 +467,27 @@ public boolean supportClientTransaction()
         return tracer;
     }
 
+    /**
+     * This helper method tells if tracing is enabled for this query. We take client provided session property
+     * as highest priority. If client does not provide any session enabling property, we then take the system
+     * default session value for determining if we should trace this query.
+     *
+     * @return
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java:475]: remove these two lines

File: presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+    public QueryStateTracingListener(Tracer tracer)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+    }
+
+    @Override
+    public void stateChanged(QueryState newState)
+    {
+        synchronized (tracer) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java:36]: Do we need to protect `previousState` as well? If yes, then we can make this method a synchronized completely

**Review by highker — APPROVED** [review]
nits only

**Comment by highker** [comment]
@tanjialiang, the test failure (`testTwoQueriesAtSameTime`) might be related. I could reproduce the issue locally. Might be a dead lock or something. Could you take a look and add a test for the state change listener?

**Comment by tanjialiang** [comment]
Fixed the test. So it turns out to be the test's problem. It uses the resource group for only 1 allowed queued query, but 2 queries were submitted at the same time, causing a possible race condition. 

The race happens when the first query is still in the queue when the second query tries to get in the queue. This case the second query will fail immediately (max queued query reached).

The race will not happen when the first query already starts running when the second query tries to get in the queue. 

So it all depends on how fast the first query gets out of the queue.

For the purpose of the test "testTwoQueriesAtSameTime" we need the max allowed queued query to be larger than 1 in order to accommodate the possible race condition.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java [hunk 23]
```diff
@@ -118,14 +120,18 @@
...
     @Inject
     public QueuedStatementResource(
             DispatchManager dispatchManager,
             DispatchExecutor executor,
             LocalQueryProvider queryResultsProvider,
             SqlParserOptions sqlParserOptions,
-            ServerConfig serverConfig)
+            ServerConfig serverConfig,
+            TracerProvider tracerProvider,
+            SessionPropertyManager sessionPropertyManager)
```
Comment (achen2289) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java:134]: how do these get initialized? not too familiar with DI, but I don't see any other changes that would update any default values for specifically the tracerProvider.
